op-node/rollup/derive/engine_controller.go

```diff
@@ -0,0 +1,224 @@
+	buildingID   eth.PayloadID
+	buildingSafe bool
+	safeAttrs    *AttributesWithParent
+}
+
+func NewEngineController(engine ExecEngine, log log.Logger, metrics Metrics, genesis rollup.Genesis) *EngineController {
+	return &EngineController{
+		engine:  engine,
+		log:     log,
+		metrics: metrics,
+		genesis: &genesis,
+	}
+}
+
+// State Getters
+
+func (e *EngineController) EngineSyncTarget() eth.L2BlockRef {
+	return e.syncTarget
+}
+
+func (e *EngineController) UnsafeL2Head() eth.L2BlockRef {
+	return e.unsafeHead
+}
+
+func (e *EngineController) PendingSafeL2Head() eth.L2BlockRef {
+	return e.pendingSafeHead
+}
+
+func (e *EngineController) SafeL2Head() eth.L2BlockRef {
+	return e.safeHead
+}
+
+func (e *EngineController) Finalized() eth.L2BlockRef {
+	return e.finalizedHead
+}
+
+func (e *EngineController) BuildingPayload() (eth.L2BlockRef, eth.PayloadID, bool) {
+	return e.buildingOnto, e.buildingID, e.buildingSafe
+}
+
+func (e *EngineController) IsEngineSyncing() bool {
+	return e.unsafeHead.Hash != e.syncTarget.Hash
+}
+
+// Setters
+
+// SetEngineSyncTarget implements LocalEngineControl.
+func (e *EngineController) SetEngineSyncTarget(r eth.L2BlockRef) {
+	e.metrics.RecordL2Ref("l2_engineSyncTarget", r)
+	e.syncTarget = r
+}
+
+// SetFinalizedHead implements LocalEngineControl.
+func (e *EngineController) SetFinalizedHead(r eth.L2BlockRef) {
+	e.metrics.RecordL2Ref("l2_finalized", r)
+	e.finalizedHead = r
+}
+
+// SetPendingSafeL2Head implements LocalEngineControl.
+func (e *EngineController) SetPendingSafeL2Head(r eth.L2BlockRef) {
+	e.metrics.RecordL2Ref("l2_pending_safe", r)
+	e.pendingSafeHead = r
+}
+
+// SetSafeHead implements LocalEngineControl.
+func (e *EngineController) SetSafeHead(r eth.L2BlockRef) {
+	e.metrics.RecordL2Ref("l2_safe", r)
+	e.safeHead = r
+}
+
+// SetUnsafeHead implements LocalEngineControl.
+func (e *EngineController) SetUnsafeHead(r eth.L2BlockRef) {
+	e.metrics.RecordL2Ref("l2_unsafe", r)
+	e.unsafeHead = r
+}
+
+// Engine Methods
+
+func (e *EngineController) StartPayload(ctx context.Context, parent eth.L2BlockRef, attrs *AttributesWithParent, updateSafe bool) (errType BlockInsertionErrType, err error) {
+	if e.IsEngineSyncing() {
+		return BlockInsertTemporaryErr, fmt.Errorf("engine is in progess of p2p sync")
+	}
+	if e.buildingID != (eth.PayloadID{}) {
+		e.log.Warn("did not finish previous block building, starting new building now", "prev_onto", e.buildingOnto, "prev_payload_id", e.buildingID, "new_onto", parent)
+		// TODO(8841): maybe worth it to force-cancel the old payload ID here.
+	}
+	fc := eth.ForkchoiceState{
+		HeadBlockHash:      parent.Hash,
+		SafeBlockHash:      e.safeHead.Hash,
+		FinalizedBlockHash: e.finalizedHead.Hash,
+	}
+
+	id, errTyp, err := startPayload(ctx, e.engine, fc, attrs.attributes)
+	if err != nil {
+		return errTyp, err
+	}
+
+	e.buildingID = id
+	e.buildingSafe = updateSafe
+	e.buildingOnto = parent
+	if updateSafe {
+		e.safeAttrs = attrs
+	}
```

Comment:
This wasn't in the old `StartPayload` from engine queue. It looks to me like the logic for when to update this is correct but it's the one noticeable logic change in this that's not just moving code around.

Given we only just started building the new payload should this variable be called `buildingSafeAttrs` or something to indicate they aren't (yet) the safe attrs? Admittedly it was `safeAttributes` in the original so probably not a big deal.

Reply:
Good callout that was an addition to match the behavior of the engine queue (we need to track if we are building a span batch or not). Ideally in a follow up we'd change the API to be like `InsertDerivedBlocked(attributes eth.(..), isSpanBatch, isLastBatchInSpan bool)`

Reply:
Ok, it's not bad as-is so happy to leave it and then hopefully the whole API will improve over time.